### PR TITLE
SQL: Fix issue with always false filter involving functions

### PR DESCRIPTION
**Pull request by matriv**
When a filter is evaluated to false then it becomes a `LocalRelation`
with an `EmptyExecutable`. The `LocalRelation` in turn, becomes a
`LocalExec` and the the `SkipQueryIfFoldingProjection` was wrongly
converting it to a `SingletonExecutable`. Moreover made a change, so
that the queries without FROM clause, which are supposed to return a
single row, to become a `LocalRelation` with a `SingletonExecutable`
instead of `EmptyExecutable` to avoid mixing up with the ones operating
on a table but with a filter that evaluates to false.

Fixes: #35980

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 1]
```diff
@@ -1849,14 +1849,18 @@ protected LogicalPlan rule(LogicalPlan plan) {
             if (plan instanceof Project) {
                 Project p = (Project) plan;
                 List<Object> values = extractConstants(p.projections());
-                if (values.size() == p.projections().size() && !(p.child() instanceof EsRelation)) {
+                if (values.size() == p.projections().size() && !(p.child() instanceof EsRelation) &&
+                    (!(p.child() instanceof LocalRelation) || (p.child() instanceof LocalRelation &&
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: The checks you added here and below are the same. Is it worth extracting these checks into a meaningfully named method and use that in both places?

**Review by astefan — APPROVED** [review]
LGTM. Left just one minor comment.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryFolderTests.java [hunk 2]
```diff
@@ -68,6 +69,38 @@ public void testFoldingToLocalExecWithProject() {
         assertThat(ee.output().get(0).toString(), startsWith("keyword{f}#"));
     }
 
+    public void testLocalExecWithPrunedFilterWithFunction() {
+        PhysicalPlan p = plan("SELECT E() FROM test WHERE PI() = 5");
```
Comment (costin) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryFolderTests.java]: Please add some tests for a local comparison that returns false as well?

**Review by costin — APPROVED** [review]
Left some small comments, otherwise LGTM.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryFolderTests.java (hunk 2 above)
Comment (matriv) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryFolderTests.java]: What do you mean local comparison?

Comment (costin) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryFolderTests.java]: `SELECT E() WHERE PI() = 5`

**Comment by matriv** [comment]
Backported to `6.x` with 29fa1a93b0addf42d04ff7da144f1c438578abe7

**Comment by matriv** [comment]
Backported to `6.5` with 075449ae0d8c7550c368e42f9869041989baf0dd

**Comment by matriv** [comment]
Backported to `6.6` with baa5e5419ab757951fa113d850bda3835536875b

**Pull request merged**